Add documentation to Microsoft.Extensions.Http

resistr opened the pull request:
Items Addressed: 

- M:Microsoft.Extensions.Http.HttpMessageHandlerBuilder.CreateHandlerPipeline
    (System.Net.Http.HttpMessageHandler,System.Collections.Generic.IEnumerable{System.Net.Http.DelegatingHandler})
- M:Microsoft.Extensions.Http.IHttpMessageHandlerBuilderFilter.Configure
    (System.Action{Microsoft.Extensions.Http.HttpMessageHandlerBuilder})

Microsoft.Extensions.Http.Logging:
- M:Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.#ctor(Microsoft.Extensions.Logging.ILogger)
- M:Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.#ctor
    (Microsoft.Extensions.Logging.ILogger,Microsoft.Extensions.Http.HttpClientFactoryOptions)
- M:Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.SendAsync
    (System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)
- M:Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.#ctor(Microsoft.Extensions.Logging.ILogger)
- M:Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.#ctor
    (Microsoft.Extensions.Logging.ILogger,Microsoft.Extensions.Http.HttpClientFactoryOptions)
- M:Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.SendAsync
    (System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)
- T:Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler
- T:Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler

Contributes to #43916

resistr committed:
Add documentation to Microsoft.Extensions.Http

- M:Microsoft.Extensions.Http.HttpMessageHandlerBuilder.CreateHandlerPipeline
    (System.Net.Http.HttpMessageHandler,System.Collections.Generic.IEnumerable{System.Net.Http.DelegatingHandler})
- M:Microsoft.Extensions.Http.IHttpMessageHandlerBuilderFilter.Configure
    (System.Action{Microsoft.Extensions.Http.HttpMessageHandlerBuilder})
- M:Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.#ctor(Microsoft.Extensions.Logging.ILogger)
- M:Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.#ctor
    (Microsoft.Extensions.Logging.ILogger,Microsoft.Extensions.Http.HttpClientFactoryOptions)
- M:Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.SendAsync
    (System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)
- M:Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.#ctor(Microsoft.Extensions.Logging.ILogger)
- M:Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.#ctor
    (Microsoft.Extensions.Logging.ILogger,Microsoft.Extensions.Http.HttpClientFactoryOptions)
- M:Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.SendAsync
    (System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)
- T:Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler
- T:Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler

Fix #43916

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=57719) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=57719) before we can accept your contribution.<br/><br/>:x: resistr [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=57719)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=57719) it.</sub>

resistr commented:
@carlossanlop Apologies I failed to add you to this as requested.

karelz commented:
@resistr thanks a lot for your PR!

I updated top post and changed "Fixes" to "Contributes to" as it does not address all items in the issue linked. I also deleted the list of APIs which are not covered by this PR from the top post. Hopefully I didn't mess up things.

MartyIX commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:
```diff
@@ -10,13 +10,20 @@
 namespace Microsoft.Extensions.Http.Logging
 {
+    /// <summary>
+    /// Handles logging of the lifecycle for a http request.
```
```suggestion
    /// Handles logging of the lifecycle for an HTTP request.
```

resistr committed:
Update src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs

Co-authored-by: Martin Vseticka <203266+MartyIX@users.noreply.github.com>

CarnaViire commented on src/libraries/Microsoft.Extensions.Http/src/HttpMessageHandlerBuilder.cs:
```diff
@@ -61,6 +61,17 @@ public abstract class HttpMessageHandlerBuilder
+        /// provided by <paramref name="additionalHandlers"/>. The resulting pipeline can be used to manually create <see cref="HttpClient"/>
+        /// or <see cref="HttpMessageInvoker"/> instances with customized message handlers.
+        /// </summary>
+        /// <param name="primaryHandler">The primary handler represents the destination of the HTTP message channel.</param>
```
I haven't heard "HTTP message channel" expression to be honest... [Our docs here](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpmessagehandler?view=net-5.0) go with "HTTP message handler" term so let's stick with it. I ended up removing the term here completely to use the type instead, but suggest to use it later in `<returns>`
```suggestion
        /// <param name="primaryHandler">An instance of <see cref="HttpMessageHandler"/> to operate at the bottom of the handler chain and actually handle the HTTP transport operations.</param>
```

CarnaViire commented on src/libraries/Microsoft.Extensions.Http/src/HttpMessageHandlerBuilder.cs:
```diff
@@ -61,6 +61,17 @@ public abstract class HttpMessageHandlerBuilder
+        /// of sending an <see cref="HttpRequestMessage"/> and receiving an <see cref="HttpResponseMessage"/>.
+        /// The handlers are invoked in a top-down fashion. That is, the first entry is invoked first for
+        /// an outbound request message but last for an inbound response message.</param>
+        /// <returns>The HTTP message channel.</returns>
```
```suggestion
        /// <returns>The HTTP message handler chain.</returns>
```

CarnaViire commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs:
```diff
@@ -10,13 +10,20 @@
 namespace Microsoft.Extensions.Http.Logging
 {
+    /// <summary>
+    /// Handles logging of the lifecycle for a http request within a scope.
```
Let's specify what the "scope" is, so it won't be confused with DI scope.
```suggestion
    /// Handles logging of the lifecycle for an HTTP request within a log scope.
```

CarnaViire reviewed (changes requested):
Thanks for the contribution! I left some comments on what I believe will be more accurate to say.

resistr committed:
Update src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs

Co-authored-by: Natalia Kondratyeva <knatalia@microsoft.com>

resistr committed:
Update src/libraries/Microsoft.Extensions.Http/src/HttpMessageHandlerBuilder.cs

Co-authored-by: Natalia Kondratyeva <knatalia@microsoft.com>

resistr committed:
Update src/libraries/Microsoft.Extensions.Http/src/HttpMessageHandlerBuilder.cs

Co-authored-by: Natalia Kondratyeva <knatalia@microsoft.com>

resistr committed:
Update to HttpMessageHandlerBuilder cstor doc

Reworded the cstor documentation on HttpMessageHandlerBuilder based on feedback from active PR.

Contributes to #43916

resistr commented on src/libraries/Microsoft.Extensions.Http/src/HttpMessageHandlerBuilder.cs:
```diff
@@ -61,6 +61,17 @@ public abstract class HttpMessageHandlerBuilder
 
+        /// <summary>
+        /// Creates an instance of an <see cref="HttpMessageHandler"/> using the <see cref="DelegatingHandler"/> instances
+        /// provided by <paramref name="additionalHandlers"/>. The resulting pipeline can be used to manually create <see cref="HttpClient"/>
```
Surry for all of the noise in this section; I had borrowed the text from https://github.com/aspnet/AspNetWebStack/blob/master/src/System.Net.Http.Formatting/HttpClientFactory.cs#L58 referenced in the method. 

I've made the recommended changes and thanks for the feedback.

resistr commented on src/libraries/Microsoft.Extensions.Http/src/HttpMessageHandlerBuilder.cs:
```diff
@@ -61,6 +61,17 @@ public abstract class HttpMessageHandlerBuilder
         public abstract HttpMessageHandler Build();
 
+        /// <summary>
+        /// Creates an instance of an <see cref="HttpMessageHandler"/> using the <see cref="DelegatingHandler"/> instances
```
I've made the recommended changes and thanks for the feedback.
@resistr

CarnaViire reviewed (approved):
LGTM, thanks!

CarnaViire commented:
@carlossanlop adding you as a reviewer as you requested in https://github.com/dotnet/runtime/issues/43916

carlossanlop commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs:
```diff
@@ -43,6 +55,7 @@ public LoggingScopeHttpMessageHandler(ILogger logger, HttpClientFactoryOptions o
             _options = options;
         }
 
+        /// <inheritdoc />
```
ditto.

resistr replied:
ditto


If not sufficient please let me know and I'll be glad to copy the inherited documentation forward explicitly.

carlossanlop commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:
```diff
@@ -27,6 +34,11 @@ public LoggingHttpMessageHandler(ILogger logger)
+        /// Initializes a new instance of the <see cref="LoggingHttpMessageHandler"/> class with a specified logger and options.
+        /// </summary>
+        /// <param name="logger">The <see cref="ILogger"/> to log to.</param>
+        /// <param name="options">The <see cref="HttpClientFactoryOptions"/> used to configure the <see cref="LoggingHttpMessageHandler"/> instance.</param>
```
```suggestion
        /// <param name="options">The <see cref="HttpClientFactoryOptions"/> used to configure the <see cref="LoggingHttpMessageHandler"/> instance.</param>
        /// <exception cref="ArgumentNullException"><paramref name="logger"/> or <paramref name="options"/> is <see langword="null"/>.</exception>
```

carlossanlop commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:
```diff
@@ -10,13 +10,20 @@
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LoggingHttpMessageHandler"/> class with a specified logger.
+        /// </summary>
+        /// <param name="logger">The <see cref="ILogger"/> to log to.</param>
```
```suggestion
        /// <param name="logger">The <see cref="ILogger"/> to log to.</param>
        /// <exception cref="ArgumentNullException"><paramref name="logger"/> is <see langword="null"/>.</exception>
```

carlossanlop commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:
```diff
@@ -43,6 +55,7 @@ public LoggingHttpMessageHandler(ILogger logger, HttpClientFactoryOptions option
             _options = options;
         }
 
+        /// <inheritdoc />
```
I'd rather have a full description for this method, especially because this SendAsync has an exception that needs to be explicitly documented.

resistr replied:
The inherited documentation includes the exception (see below). I could see being more explicit in the summary to include the fact that it is logging and sending the request; but I'm tripping on coming up with how to write that concisely. I've submitted a remark to enhance the inherited documentation.  

If not sufficient please let me know and I'll be glad to copy the inherited documentation forward explicitly.  

//
        // Summary:
        //     Sends an HTTP request to the inner handler to send to the server as an asynchronous
        //     operation.
        //
        // Parameters:
        //   request:
        //     The HTTP request message to send to the server.
        //
        //   cancellationToken:
        //     A cancellation token to cancel operation.
        //
        // Returns:
        //     The task object representing the asynchronous operation.
        //
        // Exceptions:
        //   T:System.ArgumentNullException:
        //     The request was null.

carlossanlop replied:
We can leave it as is if the description coming from the API this one inherits from, makes sense here too.
Same for the ditto API below.

carlossanlop commented on src/libraries/Microsoft.Extensions.Http/src/HttpMessageHandlerBuilder.cs:
```diff
@@ -61,6 +61,17 @@ public abstract class HttpMessageHandlerBuilder
+        /// of sending an <see cref="HttpRequestMessage"/> and receiving an <see cref="HttpResponseMessage"/>.
+        /// The handlers are invoked in a top-down fashion. That is, the first entry is invoked first for
+        /// an outbound request message but last for an inbound response message.</param>
+        /// <returns>The HTTP message handler chain.</returns>
```
```suggestion
        /// <returns>The HTTP message handler chain.</returns>
        /// <exception cref="ArgumentNullException"><paramref name="primaryHandler "/> or <paramref name="additionalHandlers "/> is <see langword="null"/>.</exception>
        /// <exception cref="InvalidOperationException"><paramref name="additionalHandlers "/> contains a <see langword="null"/> entry.
        /// -or-
        /// The <c>DelegatingHandler.InnerHandler</c> property must be <see langword="null"/>. <c>DelegatingHandler</c> instances provided to <c>HttpMessageHandlerBuilder</c> must not be reused or cached.</exception>
```

resistr committed:
Update src/libraries/Microsoft.Extensions.Http/src/HttpMessageHandlerBuilder.cs

Co-authored-by: Carlos Sanchez <1175054+carlossanlop@users.noreply.github.com>

resistr committed:
Update src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs

Co-authored-by: Carlos Sanchez <1175054+carlossanlop@users.noreply.github.com>

resistr committed:
Update src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs

Co-authored-by: Carlos Sanchez <1175054+carlossanlop@users.noreply.github.com>

resistr committed:
Update src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs

Co-authored-by: Carlos Sanchez <1175054+carlossanlop@users.noreply.github.com>

resistr committed:
Update src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs

Co-authored-by: Carlos Sanchez <1175054+carlossanlop@users.noreply.github.com>

resistr committed:
Update documentation to Microsoft.Extensions.Http

Augment documentation for SendAsync in LoggingScopeHttpMessageHandler and LoggingHttpMessageHandler

Contributes to #43916

carlossanlop reviewed (approved):
LGTM. Thank you so much for your contribution, @resistr!
